controller.go

```diff
@@ -952,29 +953,50 @@ func parseAllocateConstraintsResponse(source interface{}, machine *machine) (Con
 	if storageMatches, found := constraintsMap["storage"]; found {
 		matches := convertConstraintMatches(storageMatches)
 		for label, ids := range matches {
-			blockDevices := make([]BlockDevice, len(ids))
-			for index, id := range ids {
-				blockDevice := machine.BlockDevice(id)
-				if blockDevice == nil {
-					return empty, NewDeserializationError("constraint match storage %q: %d does not match a block device for the machine", label, id)
+			storageDevices := make([]StorageDevice, len(ids))
+			for index, storageId := range ids {
+				if id, ok := storageId.(float64); ok {
+					// Links to a block device.
+					blockDevice := machine.BlockDevice(int(id))
+					if blockDevice == nil {
+						return empty, NewDeserializationError("constraint match storage %q: %d does not match a block device for the machine", label, int(id))
+					}
+					storageDevices[index] = blockDevice
+				} else if id, ok := storageId.(string); ok {
+					// Should link to a partition.
+					const partPrefix = "partition:"
+					if strings.HasPrefix(id, partPrefix) {
```

Comment:
To save indentation, do the check for the expected prefix and exit.
Then the rest of the stuff can be done without indenting